File: frontend/src/stores/currency.ts
import { defineStore } from 'pinia'
import { getCurrenciesApi, createCurrencyApi, updateCurrencyApi, deleteCurrencyApi } from '@/api/currency'

export interface Currency {
  id: string
  company_id: string
  name: string
  symbol: string
  code: string
  is_crypto: boolean
  exchange_rate: number
  exchange_rate_op: '*' | '/'
  format_position: 'Left' | 'Right' | 'LeftSpace' | 'RightSpace'
  format_thousand_sep: string
  format_decimal_sep: string
  format_decimals: number
  is_default: boolean
}

export const useCurrencyStore = defineStore('currency', {
  state: () => ({
    currencies: [] as Currency[],
    loading: false,
    error: null as string | null
  }),

  getters: {
    getCurrencyByCode: (state) => (code: string) => {
      if (!Array.isArray(state.currencies)) return undefined
      return state.currencies.find(c => c.code === code)
    },
    defaultCurrency: (state) => {
      if (!Array.isArray(state.currencies)) return undefined
      return state.currencies.find(c => c.is_default) || state.currencies[0]
    }
  },

  actions: {
    async fetchCurrencies() {
      this.loading = true
      this.error = null
      try {
        const response = await getCurrenciesApi()
        const list = (response.data as any)?.data ?? response.data
        this.currencies = Array.isArray(list) ? list : []
      } catch (err: any) {
        this.error = err.response?.data?.error || 'Para birimleri yüklenemedi'
        this.currencies = []
      } finally {
        this.loading = false
      }
    },

    async createCurrency(payload: Partial<Currency>) {
      this.loading = true
      try {
        const response = await createCurrencyApi(payload)
        const created = (response.data as any)?.data ?? response.data
        this.currencies.push(created)
        return created
      } catch (err: any) {
        throw new Error(err.response?.data?.error || 'Para birimi eklenemedi')
      } finally {
        this.loading = false
      }
    },

    async updateCurrency(id: string, payload: Partial<Currency>) {
      this.loading = true
      try {
        const response = await updateCurrencyApi(id, payload)
        const updated = (response.data as any)?.data ?? response.data
        const index = this.currencies.findIndex(c => c.id === id)
        if (index !== -1) {
          this.currencies[index] = updated
        }
        
        // If this one is set to default, unset others locally to match backend
        if (payload.is_default) {
          this.currencies.forEach(c => {
            if (c.id !== id) c.is_default = false
          })
        }
        return updated
      } catch (err: any) {
        throw new Error(err.response?.data?.error || 'Para birimi güncellenemedi')
      } finally {
        this.loading = false
      }
    },

    async deleteCurrency(id: string) {
      this.loading = true
      try {
        await deleteCurrencyApi(id)
        this.currencies = this.currencies.filter(c => c.id !== id)
      } catch (err: any) {
        throw new Error(err.response?.data?.error || 'Para birimi silinemedi')
      } finally {
        this.loading = false
      }
    }
  }
})

File: backend/models/currency.go
package models

import (
	"time"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"
	"gorm.io/gorm"
)

type Currency struct {
	ID                 uuid.UUID       `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	CompanyID          uuid.UUID       `gorm:"type:uuid;not null;index" json:"company_id"`
	Name               string          `gorm:"type:varchar(100);not null" json:"name"`
	Symbol             string          `gorm:"type:varchar(10);not null" json:"symbol"`
	Code               string          `gorm:"type:varchar(10);not null" json:"code"`
	IsCrypto           bool            `gorm:"default:false" json:"is_crypto"`
	ExchangeRate       decimal.Decimal `gorm:"type:numeric(65,4);default:1.0" json:"exchange_rate"`
	ExchangeRateOp     string          `gorm:"type:varchar(5);default:'*'" json:"exchange_rate_op"` // '*' veya '/'
	FormatPosition     string          `gorm:"type:varchar(20);default:'Left'" json:"format_position"` // Left, Right, LeftSpace, RightSpace
	FormatThousandSep  string          `gorm:"type:varchar(5);default:','" json:"format_thousand_sep"`
	FormatDecimalSep   string          `gorm:"type:varchar(5);default:'.'" json:"format_decimal_sep"`
	FormatDecimals     int             `gorm:"default:2" json:"format_decimals"`
	IsDefault          bool            `gorm:"default:false" json:"is_default"`

	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`
}
